client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java

```diff
@@ -135,22 +143,55 @@ private boolean isMemoryShuffleEnabled(String storageType) {
   public void write(Iterator<Product2<K, V>> records) throws IOException {
     List<ShuffleBlockInfo> shuffleBlockInfos = null;
     Set<Long> blockIds = Sets.newConcurrentHashSet();
+
+    boolean isCombine = shuffleDependency.mapSideCombine();
+    SizeTrackingAppendOnlyMap<K, V> sizeTrackingAppendOnlyMap = null;
+    if (isCombine) {
+      sizeTrackingAppendOnlyMap = new SizeTrackingAppendOnlyMap<>();
+    }
+
     while (records.hasNext()) {
       Product2<K, V> record = records.next();
-      K key = record._1();
-      int partition = getPartition(key);
-      boolean isCombine = shuffleDependency.mapSideCombine();
       if (isCombine) {
-        Function1 createCombiner = shuffleDependency.aggregator().get().createCombiner();
-        Object c = createCombiner.apply(record._2());
-        shuffleBlockInfos = bufferManager.addRecord(partition, record._1(), c);
+        final Function1 createCombiner = shuffleDependency.aggregator().get().createCombiner();
+        if (isMergeEnabled) {
+          final Function2 mergeValues = shuffleDependency.aggregator().get().mergeValue();
+          Function2 updateFunc = new AbstractFunction2<Boolean, V, V>() {
+            @Override
+            public V apply(Boolean hasKey, V oldVal) {
+              if (hasKey) {
+                return (V) mergeValues.apply(oldVal, record._2());
+              }
+              return (V) createCombiner.apply(record._2());
+            }
+          };
+          sizeTrackingAppendOnlyMap.changeValue(record._1(), updateFunc);
+
+          if (sizeTrackingAppendOnlyMap.estimateSize() > 1 << 15) {
+            shuffleBlockInfos = toBufferManager(sizeTrackingAppendOnlyMap);
+            sizeTrackingAppendOnlyMap = new SizeTrackingAppendOnlyMap<>();
+          }
+        } else {
+          int partition = getPartition(record._1());
+          shuffleBlockInfos = bufferManager.addRecord(partition, record._1(), createCombiner.apply(record._2()));
+        }
       } else {
+        int partition = getPartition(record._1());
         shuffleBlockInfos = bufferManager.addRecord(partition, record._1(), record._2());
       }
       if (shuffleBlockInfos != null && !shuffleBlockInfos.isEmpty()) {
         processShuffleBlockInfos(shuffleBlockInfos, blockIds);
       }
     }
+
+    // If merge enable, add all the rest records from map to buffer manager.
+    if (sizeTrackingAppendOnlyMap != null && !sizeTrackingAppendOnlyMap.isEmpty()) {
```

Comment:
If we use `SizeTrackingAppendOnlyMap<K, V> sizeTrackingAppendOnlyMap = new SizeTrackingAppendOnlyMap<>();` Here we can simplify it 
`!sizeTrackingAppendOnlyMap.isEmpty()`